core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1521 @@
+   */
+  def curReassignmentsToString(adminClient: Admin): String = {
+    val currentReassignments = adminClient.
+      listPartitionReassignments().reassignments().get().asScala
+    val text = currentReassignments.keySet.toList.sortWith(compareTopicPartitions).map {
+      case part =>
+        val reassignment = currentReassignments(part)
+        val replicas = reassignment.replicas().asScala
+        val addingReplicas = reassignment.addingReplicas().asScala
+        val removingReplicas = reassignment.removingReplicas().asScala
+        "%s: replicas: %s.%s%s".format(part, replicas.mkString(","),
+          if (addingReplicas.isEmpty) "" else
+            " adding: %s.".format(addingReplicas.mkString(",")),
+          if (removingReplicas.isEmpty) "" else
+            " removing: %s.".format(removingReplicas.mkString(",")))
+    }.mkString(System.lineSeparator())
+    if (text.isEmpty) {
+      "No partition reassignments found."
     } else {
-      printCurrentAssignment(zkClient, partitionAssignment.map(_._1.topic))
-      if (throttle.interBrokerLimit >= 0 || throttle.replicaAlterLogDirsLimit >= 0)
-        println(String.format("Warning: You must run Verify periodically, until the reassignment completes, to ensure the throttle is removed. You can also alter the throttle by rerunning the Execute command passing a new value."))
-      if (reassignPartitionsCommand.reassignPartitions(throttle, timeoutMs)) {
-        println("Successfully started reassignment of partitions.")
-      } else
-        println("Failed to reassign partitions %s".format(partitionAssignment))
+      "Current partition reassignments:%n%s".format(text)
+    }
+  }
+
+  /**
+   * Verify that all the brokers in an assignment exist.
+   *
+   * @param adminClient                 The AdminClient to use.
+   * @param brokers                     The broker IDs to verify.
+   */
+  def verifyBrokerIds(adminClient: Admin, brokers: Set[Int]): Unit = {
+    val allNodeIds = adminClient.describeCluster().nodes().get().asScala.map(_.id).toSet
+    brokers.find(!allNodeIds.contains(_)).map {
+      case id => throw new AdminCommandFailedException(s"Unknown broker id ${id}")
+    }
+  }
+
+  /**
+   * The entry point for the --execute command.
+   *
+   * @param zkClient                    The ZooKeeper client to use.
+   * @param reassignmentJson            The JSON string to use for the topics to reassign.
+   * @param interBrokerThrottle         The inter-broker throttle to use, or a negative number
+   *                                    to skip using a throttle.
+   */
+  def executeAssignment(zkClient: KafkaZkClient,
+                        reassignmentJson: String,
+                        interBrokerThrottle: Long): Unit = {
+    val (proposedParts, proposedReplicas) = parseExecuteAssignmentArgs(reassignmentJson)
+    if (!proposedReplicas.isEmpty) {
+      throw new AdminCommandFailedException("bootstrap-server needs to be provided when " +
```

Comment:
Do we have a test to ensure that the old way to run replica log dir reassignments with ZK works? i.e `./kafka-reassign-partitions --zookeeper ZK --bootstrap-server server --reassignment-json-file file_with_dir_reassignments.json`
From reading the code, I think so - but always worth making sure

Reply:
If you supply both `--zookeeper` and `--bootstrap-server`, then --zookeeper is ignored.  So there's no need for separate tests for the combination.

Reply:
It would be nice if we gave a warning that the `--zookeeper` flag was being ignore in this case